tag subcommand fails with unhelpful error message when the tag exists

Running `dune-release tag` fails if the tag already exists:
```
$ dune-release tag
[-] Extracting tag from first entry in CHANGES.md
[-] Using tag "v0.2.0+test"
[?] Create git tag v0.2.0+test for HEAD? [Y/n]

dune-release: [ERROR] The following command exited with code 128:
                      git --git-dir .git tag -a -m Distribution v0.2.0+test
                      v0.2.0+test HEAD
```

This isn't very helpful. I think that to the very least we should:
1. clean up the error logs in those cases, I'm really not so fond of `Logs` default error reporting.
2. Show external commands stderr/stdout when they fail

In this particular case we could go even further and check whether the tag exists before tagging.If it exists and points to the current `HEAD`, we could successfully exit, informing the user that the tag already exists and points to the right place. If it points to somewhere else, we could fail and inform the user why we did.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -10,6 +10,9 @@
 
 ### Changed
 
+- Command `tag`: improve error and log messages by comparing the provided 
+  commit with the commit correspondent to the provided tag (#226, @pitag-ha)
+
 ### Deprecated
 
 - Deprecate opam 1.x (#195, @gpetiot)

--- a/bin/tag.ml
+++ b/bin/tag.ml
@@ -8,20 +8,52 @@ open Bos_setup
 open Dune_release
 
 let vcs_tag tag ~dry_run ~commit_ish ~force ~sign ~delete ~msg ~yes =
-  let msg = match msg with None -> strf "Distribution %s" tag | Some m -> m in
   Vcs.get () >>= fun repo ->
-  match delete with
-  | true ->
-      Prompt.confirm_or_abort ~yes ~question:(fun l ->
-          l "Delete tag %a?" Text.Pp.version tag)
-      >>= fun () ->
+  Vcs.commit_id ~dirty:false ~commit_ish repo
+  |> R.reword_error (fun (`Msg msg) ->
+         R.msgf "Invalid commit-ish %s: %s" commit_ish msg)
+  >>= fun commit ->
+  let tag_commit_opt = Vcs.tag_points_to ~tag repo in
+  match (tag_commit_opt, delete) with
+  | Some tag_commit, true ->
+      let question =
+        if tag_commit = commit then
+          Prompt.confirm_or_abort ~yes ~question:(fun l ->
+              l "Delete tag %a?" Text.Pp.version tag)
+        else
+          Prompt.confirm_or_abort_neg ~yes ~question:(fun l ->
+              l
+                "%a Tag %a does not point to the commit you've provided \
+                 (default: HEAD). Do you want to delete it anyways?"
+                Fmt.(styled `Red string)
+                "Warning:" Text.Pp.version tag)
+      in
+      question >>= fun () ->
       Vcs.delete_tag ~dry_run repo tag >>| fun () ->
       App_log.success (fun m -> m "Deleted tag %a" Text.Pp.version tag)
-  | false ->
+  | None, true ->
+      Ok
+        (App_log.status (fun apply_log_l ->
+             apply_log_l "Nothing to be deleted: there is no tag %a."
+               Text.Pp.version tag))
+  | Some tag_commit, false ->
+      if tag_commit = commit then
+        Ok
+          (App_log.status (fun apply_log_l ->
+               apply_log_l "Nothing to be done: tag already exists."))
+      else
+        R.error_msgf
+          "A tag with name %a already exists, but points to a different \
+           commit. You can delete that tag using the `-d` flag."
+          Text.Pp.version tag
+  | None, false ->
       Prompt.confirm_or_abort ~yes ~question:(fun l ->
           l "Create git tag %a for %a?" Text.Pp.version tag Text.Pp.commit
             commit_ish)
       >>= fun () ->
+      let msg =
+        match msg with None -> strf "Distribution %s" tag | Some m -> m
+      in
       Vcs.tag repo ~dry_run ~force ~sign ~msg ~commit_ish tag >>| fun () ->
       App_log.success (fun m ->
           m "Tagged %a with version %a" Text.Pp.commit commit_ish

--- a/lib/prompt.ml
+++ b/lib/prompt.ml
@@ -1,22 +1,34 @@
-let ask f =
+type default = Yes | No
+
+let ask f ~default_answer =
+  let options : ('a, Format.formatter, unit, unit) format4 =
+    match default_answer with Yes -> " [Y/n]" | No -> " [y/N]"
+  in
   App_log.question (fun l ->
-      f (fun ?header ?tags fmt -> l ?header ?tags (fmt ^^ " [Y/n]")))
+      f (fun ?header ?tags fmt -> l ?header ?tags (fmt ^^ options)))
+
+let rec loop ~question ~default_answer =
+  ask question ~default_answer;
+  match String.lowercase_ascii (read_line ()) with
+  | "" when default_answer = Yes -> true
+  | "" when default_answer = No -> false
+  | "y" | "yes" -> true
+  | "n" | "no" -> false
+  | _ ->
+      App_log.unhappy (fun l ->
+          l
+            "Please answer with \"y\" for yes, \"n\" for no or just hit enter \
+             for the default");
+      loop ~question ~default_answer
 
 let confirm ~question ~yes =
-  let rec loop () =
-    ask question;
-    match String.lowercase_ascii (read_line ()) with
-    | "" | "y" | "yes" -> true
-    | "n" | "no" -> false
-    | _ ->
-        App_log.unhappy (fun l ->
-            l
-              "Please answer with \"y\" for yes, \"n\" for no or just hit \
-               enter for the default");
-        loop ()
-  in
-  if yes then true else loop ()
+  if yes then true else loop ~question ~default_answer:Yes
 
 let confirm_or_abort ~question ~yes =
   if confirm ~question ~yes then Ok ()
+  else Error (`Msg "Aborting on user demand")
+
+let confirm_or_abort_neg ~question ~yes =
+  if yes then Ok ()
+  else if loop ~question ~default_answer:No then Ok ()
   else Error (`Msg "Aborting on user demand")

--- a/lib/prompt.mli
+++ b/lib/prompt.mli
@@ -11,3 +11,7 @@ val confirm_or_abort :
   question:('a, unit) Logs.msgf -> yes:bool -> (unit, Rresult.R.msg) result
 (** Same as [confirm] but returns [Ok ()] for yes and
     [Error (`Msg "Aborting on user demand")] for no *)
+
+val confirm_or_abort_neg :
+  question:('a, unit) Logs.msgf -> yes:bool -> (unit, Rresult.R.msg) result
+(** Like [confirm_or_abort], but defaults to no when hitting enter *)

--- a/lib/vcs.ml
+++ b/lib/vcs.ml
@@ -400,6 +400,9 @@ let tag_exists ~dry_run r tag =
   | (`Git, _, _) as r -> git_tag_exists r ~dry_run tag
   | `Hg, _, _ -> failwith "TODO"
 
+let tag_points_to r ~tag =
+  commit_id ~dirty:false ~commit_ish:tag r |> R.to_option
+
 let branch_exists ~dry_run r tag =
   match r with
   | (`Git, _, _) as r -> git_branch_exists r ~dry_run tag

--- a/lib/vcs.mli
+++ b/lib/vcs.mli
@@ -93,6 +93,8 @@ val tags : t -> (string list, R.msg) result
 
 val tag_exists : dry_run:bool -> t -> string -> bool
 
+val tag_points_to : t -> tag:string -> string option
+
 val branch_exists : dry_run:bool -> t -> string -> bool
 
 val changes :

--- a/tests/bin/tag/dune
+++ b/tests/bin/tag/dune
@@ -1,0 +1,3 @@
+(mdx
+ (files run.t)
+ (packages dune-release))

--- a/tests/bin/tag/run.t
+++ b/tests/bin/tag/run.t
@@ -1,0 +1,74 @@
+We need a basic opam project skeleton
+
+    $ cat > CHANGES.md << EOF \
+    > ## 0.1.0\
+    > \
+    > - Some other feature\
+    > \
+    > ## 0.0.0\
+    > \
+    > - Some feature\
+    > EOF
+    $ touch whatever.opam
+    $ cat > dune-project << EOF \
+    > (lang dune 2.4)\
+    > (name whatever)\
+    > EOF
+
+We need to set up a git project with two commits to test trying to tag different commits with the same tag name.
+
+    $ git init > /dev/null
+    $ git add whatever.opam dune-project
+    $ git commit -m "Initial commit" > /dev/null
+    $ git add CHANGES.md
+    $ git commit -m "Add CHANGES.md" > /dev/null
+
+Running `dune-release tag` for the first time should tag HEAD with the current version number.
+
+    $ dune-release tag -y
+    [-] Extracting tag from first entry in CHANGES.md
+    [-] Using tag "0.1.0"
+    [+] Tagged HEAD with version 0.1.0
+
+Running `dune-release tag` again should inform the user that that tag already exists.
+
+    $ dune-release tag
+    [-] Extracting tag from first entry in CHANGES.md
+    [-] Using tag "0.1.0"
+    [-] Nothing to be done: tag already exists.
+
+Running `dune-release tag` again, but providing a different commit should inform the user that that tag already exists but points to a different commit.
+
+    $ dune-release tag --commit=HEAD^
+    [-] Extracting tag from first entry in CHANGES.md
+    [-] Using tag "0.1.0"
+    dune-release: [ERROR] A tag with name 0.1.0 already exists, but points to a
+                          different commit. You can delete that tag using the
+                          `-d` flag.
+    [3]
+
+Trying to delete the created tag providing a different commit should give a warning. The answer to the question 
+asking for confirmation should default to "no".
+
+    $ echo "" | dune-release tag -d --commit=HEAD^
+    [-] Extracting tag from first entry in CHANGES.md
+    [-] Using tag "0.1.0"
+    [?] Warning: Tag 0.1.0 does not point to the commit you've provided (default: HEAD). Do you want to delete it anyways? [y/N]
+    dune-release: [ERROR] Aborting on user demand
+    [3]
+
+Deleting the created tag providing the commit it points to (here the default, so HEAD) should work without warning. 
+The answer to the question asking for confirmation should default to "yes".
+
+    $ echo "" | dune-release tag -d
+    [-] Extracting tag from first entry in CHANGES.md
+    [-] Using tag "0.1.0"
+    [?] Delete tag 0.1.0? [Y/n]
+    [+] Deleted tag 0.1.0
+
+Trying to delete a commit that doesn't exist should inform the user that there's nothing to be deleted.
+
+    $ dune-release tag -d -y
+    [-] Extracting tag from first entry in CHANGES.md
+    [-] Using tag "0.1.0"
+    [-] Nothing to be deleted: there is no tag 0.1.0.

